File: SnapBackApp/Sources/SnapBackApp/Bridge/NonceCache.swift
import Foundation

/// Bounded TTL+LRU cache for replay protection.
/// Thread-safe via an internal serial queue.
public final class NonceCache {
    private let capacity: Int
    private let ttl: TimeInterval
    private var entries: [(nonce: String, insertedAt: TimeInterval)] = []
    private let queue = DispatchQueue(label: "com.snapback.bridge.nonceCache")

    public init(capacity: Int, ttlSeconds: TimeInterval) {
        self.capacity = capacity
        self.ttl = ttlSeconds
    }

    /// Returns `true` if the nonce was new and has now been recorded.
    /// `now` is expressed in unix seconds; the caller supplies it so tests are deterministic.
    public func tryAdd(_ nonce: String, at now: TimeInterval) -> Bool {
        queue.sync {
            entries.removeAll { now - $0.insertedAt > ttl }
            if entries.contains(where: { $0.nonce == nonce }) { return false }
            entries.append((nonce: nonce, insertedAt: now))
            while entries.count > capacity { entries.removeFirst() }
            return true
        }
    }
}

File: SnapBackMobile/app/src/main/kotlin/com/snapback/mobile/protocol/NonceCache.kt
package com.snapback.mobile.protocol

class NonceCache(private val capacity: Int, private val ttlSeconds: Double) {
    private data class Entry(val nonce: String, val insertedAt: Double)
    private val entries = ArrayDeque<Entry>()

    @Synchronized
    fun tryAdd(nonce: String, at: Double): Boolean {
        while (entries.isNotEmpty() && at - entries.first().insertedAt > ttlSeconds) {
            entries.removeFirst()
        }
        if (entries.any { it.nonce == nonce }) return false
        entries.addLast(Entry(nonce, at))
        while (entries.size > capacity) entries.removeFirst()
        return true
    }
}
